---
title: "Dataset Intersections"
output: html_document
date: "2024-12-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

2024/12/02: Now unlike the way I did it in the big combined script, here I am going to try out doing more stringent FDR cutoffs as well as fairly relaxed log2FC cutoffs. I may also change the methylation list input, in fact I probably should (either use limma with cbioportal or confirm results from the probe-wise version).

2024/12/05: Some thoughts I have:

- If I use quite stringent criteria (FDR cutoffs of 0.001 and log2FC cutoffs of 0.5, 1.5, and 0.5 respectively), I get 29 "linked genes" which have an interesting interaction network in STRING. (I picked the log2FC cutoffs by checking the histograms of log2FC values and trying to find cutoffs that left normally distributed-looking tails.)

- If I use the most relaxed criteria possible (FDR cutoffs of 0.05 and log2FC cutoffs of 0) I get 27 "persistent linked" genes which have a somewhat interesting interaction network in STRING.

- If I want to do pathway analysis and gene-list comparisons (comparing with outputs from other studies), it seems I will need a bigger gene list. So I will have to use "linked genes" with less stringent criteria, and not "persistent linked", because these are narrowed down to 27 with even the least stringent criteria. If I try this with the most relaxed criteria possible (FDR cutoffs of 0.05 and log2FC cutoffs of 0), I get 747 linked genes, which yield 8 significant pathways in GO.

- If I am looking at meta-analysis comparisons across airway datasets, as well as survival curves, maybe I would be most interested in using the list of 27 "persistent linked" genes because we would want to back up the idea that these are definitely differentially expressed in smoker airways.

- 2025/01/29: Now I am using the versions of TE and TM that had never smoker samples removed from them

- 2025/01/31: This version uses the LUSC data instead of LUAD data.

# Comparing and filtering A1, TE, TM, and A2 to generate "linked genes" and "linked persistent genes" lists

## Reading in datasets to intersect
```{r}

# A1
GSE63127_CS_NS              <- read.table("../2_Outputs/1_Airway_DEGs/GSE63127_CS_NS_limma_20241204.txt", header = TRUE)

# T-E
TCGA_LUSC_DEG               <- read.table("../2_Outputs/4_Tumor_DEGs/TCGA_LUSC_limma_DEG_20250130.txt")

# T-M
TCGA_LUSC_DMG               <- read.table("../2_Outputs/5_Tumor_DMGs/TCGA_LUSC_limma_DMG_20250131.txt")

# A2
GSE7895_persistent_DEGs     <- read.table("../2_Outputs/1_Airway_DEGs/GSE7895_persistent_DEGs_20241127.txt")

```

#######

# I. Using more stringent FDR cutoffs

## I.i Testing FDR and log2FC cutoffs
```{r}
library(dplyr)
### Testing FDR and log2FC cutoffs ###

# A1 DEGs
FDR_cutoff_A1 = 0.01
log2FC_cutoff_A1 = 0.25

print(paste0("A1 DEGs meeting FDR cutoff ", FDR_cutoff_A1, " and log2FC cutoff ", log2FC_cutoff_A1))
nrow(GSE63127_CS_NS[GSE63127_CS_NS$adj.P.Val<FDR_cutoff_A1
                                    & abs(GSE63127_CS_NS$logFC)>log2FC_cutoff_A1,])

hist(GSE63127_CS_NS[GSE63127_CS_NS$adj.P.Val<FDR_cutoff_A1
                                    & abs(GSE63127_CS_NS$logFC)>log2FC_cutoff_A1,]$logFC, breaks = 100)

hist(log10(GSE63127_CS_NS[GSE63127_CS_NS$adj.P.Val<FDR_cutoff_A1
                                    & abs(GSE63127_CS_NS$logFC)>log2FC_cutoff_A1,]$adj.P.Val), breaks = 100)

# TCGA-LUSC DEGs
FDR_cutoff_TE = 0.001
log2FC_cutoff_TE = 0.75

print(paste0("TE DEGs meeting FDR cutoff ", FDR_cutoff_TE, " and log2FC cutoff ", log2FC_cutoff_TE))
nrow(TCGA_LUSC_DEG[TCGA_LUSC_DEG$FDR<FDR_cutoff_TE
                              & abs(TCGA_LUSC_DEG$log2FC)>log2FC_cutoff_TE,])

hist(TCGA_LUSC_DEG[TCGA_LUSC_DEG$FDR<FDR_cutoff_TE
                              & abs(TCGA_LUSC_DEG$log2FC)>log2FC_cutoff_TE,]$log2FC, breaks = 100)

hist(log10(TCGA_LUSC_DEG[TCGA_LUSC_DEG$FDR<FDR_cutoff_TE
                              & abs(TCGA_LUSC_DEG$log2FC)>log2FC_cutoff_TE,]$FDR), breaks = 100)


# TCGA-LUSC DMGs limma
FDR_cutoff_TM = 0.01
log2FC_cutoff_TM = 0.25

print(paste0("TM DMGs meeting FDR cutoff ", FDR_cutoff_TM, " and log2FC cutoff ", log2FC_cutoff_TM))
nrow(TCGA_LUSC_DMG[TCGA_LUSC_DMG$FDR<FDR_cutoff_TM
                               & abs(TCGA_LUSC_DMG$log2FC)>log2FC_cutoff_TM,])

hist(TCGA_LUSC_DMG[TCGA_LUSC_DMG$FDR<FDR_cutoff_TM
                               & abs(TCGA_LUSC_DMG$log2FC)>log2FC_cutoff_TM,]$log2FC, breaks = 100)

hist(log10(TCGA_LUSC_DMG[TCGA_LUSC_DMG$FDR<FDR_cutoff_TM
                               & abs(TCGA_LUSC_DMG$log2FC)>log2FC_cutoff_TM,]$FDR), breaks = 100)


```

### Applying cutoffs and formatting consistently for merge ###
```{r}
GSE63127_CS_NS_cutoff <- GSE63127_CS_NS %>%
  filter(abs(logFC)>log2FC_cutoff_A1) %>%
  filter(adj.P.Val < FDR_cutoff_A1) %>%
  dplyr::rename(., Gene = Gene.symbol, FDR_A1 = adj.P.Val, log2FC_A1 = logFC) %>%
  dplyr::select(., Gene, FDR_A1, log2FC_A1)

TCGA_LUSC_DEG_cutoff <- TCGA_LUSC_DEG %>%
  filter(abs(log2FC)>log2FC_cutoff_TE) %>%
  filter(FDR < FDR_cutoff_TE) %>%
  dplyr::rename(., FDR_TE = FDR, log2FC_TE = log2FC, Gene = Gene) %>%
  dplyr::select(., Gene, FDR_TE, log2FC_TE)

TCGA_LUSC_DMG_cutoff <- TCGA_LUSC_DMG %>%
  filter(abs(log2FC)>log2FC_cutoff_TM) %>%
  filter(FDR < FDR_cutoff_TM) %>%
  dplyr::rename(., FDR_TM = FDR, log2FC_TM = log2FC) %>%
  dplyr::select(., Gene, FDR_TM, log2FC_TM)
```


### I.ii Intersection of A1 and TE DEGs
```{r}
## Merge the lists

A1_TE_merged_DEGs <- GSE63127_CS_NS_cutoff %>%
  inner_join(., TCGA_LUSC_DEG_cutoff, by = "Gene") %>%
  filter(sign(log2FC_A1)==sign(log2FC_TE)) # Filter to genes with same signs

print(c("overlap", nrow(A1_TE_merged_DEGs)))

## Fisher's exact test (In progress)
n1 <- nrow(A1_TE_merged_DEGs)
n2 <- nrow(TCGA_LUSC_DEG_cutoff)
n3 <- nrow(GSE63127_CS_NS_cutoff)

print(c("expected overlap ", n2*n3/20000))

fisher.test(matrix(c(n1, n3-n1, n2-n1, 20000-n3-n2+n1), nrow = 2) )


```

### I.iii Intersection of A1/TE DEGs with TM DMGs
```{r}
## Merge the lists

A1_TE_TM_linked_genes <- A1_TE_merged_DEGs %>%
  inner_join(., TCGA_LUSC_DMG_cutoff, by = "Gene") %>%
  filter(sign(log2FC_A1)!=sign(log2FC_TM)) # Filter to genes with opposite sign

nrow(A1_TE_TM_linked_genes)

#write(A1_TE_TM_linked_genes$Gene, "../2_Outputs/6_Linked_genes/A1_TE_TM_linked_genes_LUSC_0.01_0.25_0.001_0.75_0.01_0.25_20250206.txt", sep = "\n")
```

### Fisher's exact test for significance of overlaps

```{r}

## Expected overlap given independent datasets:
print("Expected overlap:")

n4<-n1*n2
n5 <- n3/20000^2
n4*n5

## Fisher's exact test (In progress - should not just use 20k, that is a stand-in for all genes)
## Also need to account for overlaps better in the third term there - would need to check other overlaps too

# Known values
overlap <- 91 # Observed overlap
universe_size <- 20000 # 20k genes in genome
union_size <- length(unique(c(GSE63127_CS_NS_cutoff$Gene, GSE63127_CS_NS_cutoff$Gene, TCGA_LUSC_DMG_cutoff$Gene))) # Size of |A ∪ B ∪ C|

# Calculate table entries
n11 <- overlap
n12 <- union_size - overlap
n21 <- universe_size - union_size
n22 <- 0 # Genes outside all sets

# Construct 2x2 contingency table
contingency_table <- matrix(c(n11, n12, n21, n22), nrow = 2)

# Perform Fisher's Exact Test
fisher_result <- fisher.test(contingency_table)

# Output results
fisher_result


```


### I.iv.1 Visualizing Spearman correlations of log2FC values
```{r}
## Visualizing Spearman correlations
library(GGally)
ggpairs(A1_TE_TM_linked_genes[, c("log2FC_A1", "log2FC_TE", "log2FC_TM")],
  upper = list(continuous = wrap("cor", method = "spearman")),
  lower = list(continuous = "points", combo = "facethist", discrete = "facetbar", na ="na"),
  diag = list(continuous = "densityDiag", discrete = "barDiag", na = "naDiag"))

```

### I.iv.2 Visualizing multiple linear regression of the log2FC values
```{r}
mlr_airway_model <- lm(log2FC_A1 ~ log2FC_TE + log2FC_TM, data = A1_TE_TM_linked_genes)
summary(mlr_airway_model)


library(scatterplot3d)
s3d <- scatterplot3d(x = A1_TE_TM_linked_genes$log2FC_A1, 
              y = A1_TE_TM_linked_genes$log2FC_TE, 
              z = A1_TE_TM_linked_genes$log2FC_TM,
              main="Plotting airway expression vs. tumor expression vs. tumor methylation",
              xlab = "log2FC(TE)",
              ylab = "log2FC(TM)",
              zlab = "log2FC(A1)",
              pch = 19,
              #color = colors_3d, # getting color values from methylation sign
              color = "steelblue",
              labels = A1_TE_TM_linked_genes$Gene,
              type = "h",
              #highlight.3d = TRUE, 
              angle = 60
              )
s3d$plane3d(mlr_airway_model) # Add a plane based on the multiple linear regression model

```

OK here is an amazing interaction network I got:

https://version-12-0.string-db.org/cgi/network?networkId=bK3IJNEMHJEy

This was using the 29 linked genes given by the parameters:

FDR_cutoff_A1=0.001
log2FC_cutoff_A1=0.5

FDR_cutoff_TE=0.001
log2FC_cutoff_TE = 1.5

FDR_cutoff_TM=0.001
log2FC_cutoff_TM = 0.5

Now to check via intersection with the "persistent" genes:


## 5.3 Comparing and filtering A2 and "linked" genes to give "persistent linked genes"

### Filtering by dataset intersection
```{r}

A1_TE_TM_A2_persistent_linked_genes <- A1_TE_TM_linked_genes %>%
  dplyr::inner_join(., GSE7895_persistent_DEGs, by = "Gene") %>%
  dplyr::filter(sign(log2FC_A1) == sign(CS_NS_A2)) # Ensure signs are the same between A1 and A2 CS vs NS

nrow(A1_TE_TM_A2_persistent_linked_genes)

A1_TE_TM_A2_persistent_linked_genes$Gene

```

### 5.3.3 Additional checks (correlation of log2FCs in A1 and A2)
```{r}
cor.test(A1_TE_TM_A2_persistent_linked_genes$log2FC_A1, y = A1_TE_TM_A2_persistent_linked_genes$CS_NS_A2, use = "everything",
    method = "spearman")
    
ggplot(A1_TE_TM_A2_persistent_linked_genes, aes(x = log2FC_A1, y = CS_NS_A2)) +
  geom_point(color = "blue", size = 3) +                # Scatter points
  geom_smooth(method = "lm", se = FALSE, color = "black") + # Add a trend line +
  theme_minimal()
```







## 2025/02/07: This cutoff optimization code is adapted from the code I did in the version of this script from 20241210_12PM

## Attempting to optimize cutoffs to satisfy Fisher's exact test while maintaining acceptable numbers of DEGs

### Making functions to do dataset intersections and get number of resulting rows
```{r}

linked_gene_result <- function(GSE63127_CS_NS, TCGA_LUSC_DEG, TCGA_LUSC_DMG, 
                               log2FC_cutoff_A1, FDR_cutoff_A1,
                               log2FC_cutoff_TE, FDR_cutoff_TE,
                               log2FC_cutoff_TM, FDR_cutoff_TM){
  
  GSE63127_CS_NS_cutoff <- GSE63127_CS_NS %>%
    filter(abs(logFC)>log2FC_cutoff_A1) %>%
    filter(adj.P.Val < FDR_cutoff_A1) %>%
    dplyr::rename(., Gene = Gene.symbol, FDR_A1 = adj.P.Val, log2FC_A1 = logFC) %>%
    dplyr::select(., Gene, FDR_A1, log2FC_A1)

  TCGA_LUSC_DEG_cutoff <- TCGA_LUSC_DEG %>%
    filter(abs(log2FC)>log2FC_cutoff_TE) %>%
    filter(FDR < FDR_cutoff_TE) %>%
    dplyr::rename(., FDR_TE = FDR, log2FC_TE = log2FC) %>%
    dplyr::select(., Gene, FDR_TE, log2FC_TE)

  TCGA_LUSC_DMG_cutoff <- TCGA_LUSC_DMG %>%
    filter(abs(log2FC)>log2FC_cutoff_TM) %>%
    filter(FDR < FDR_cutoff_TM) %>%
    dplyr::rename(., FDR_TM = FDR, log2FC_TM = log2FC) %>%
    dplyr::select(., Gene, FDR_TM, log2FC_TM)
  
  A1_TE_merged_DEGs <- GSE63127_CS_NS_cutoff %>%
    inner_join(., TCGA_LUSC_DEG_cutoff, by = "Gene") %>%
    filter(sign(log2FC_A1)==sign(log2FC_TE)) # Filter to genes with same signs
  
  A1_TE_TM_linked_genes <- A1_TE_merged_DEGs %>%
    inner_join(., TCGA_LUSC_DMG_cutoff, by = "Gene") %>%
    filter(sign(log2FC_A1)!=sign(log2FC_TM)) # Filter to genes with opposite sign
  
  return(nrow(A1_TE_TM_linked_genes))

}
```


## Trying many combinations of cutoff values and observing effect on DEGs count and Fisher's test
```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Define ranges for cutoffs
fdr_cutoffs_A1 <- c(0.01, 0.05)
fdr_cutoffs_TE <- c(0.001, 0.01, 0.05)
fdr_cutoffs_TM <- c(0.01, 0.05)

log2fc_cutoffs_A1 <- c(0.25, 0.5)
log2fc_cutoffs_TE <- c(0.25, 0.75, 1)
log2fc_cutoffs_TM <- c(0.25, 0.5)

# Create all combinations of cutoffs for A1, TE, and TM
cutoff_combinations <- expand.grid(FDR_A1 = fdr_cutoffs_A1,
                                    FDR_TE = fdr_cutoffs_TE,
                                    FDR_TM = fdr_cutoffs_TM,
                                    log2FC_A1 = log2fc_cutoffs_A1,
                                    log2FC_TE = log2fc_cutoffs_TE,
                                    log2FC_TM = log2fc_cutoffs_TM)

# Calculate rows removed for each combination, remaining DEGs in each dataset after cutoffs are applied, number of linked genes, expected overlap assuming 20000 genes, and fisher's exact test result
A1_DEGs <- integer()
TE_DEGs <- integer()
TM_DMGs <- integer()
linked_genes <- integer()
expected_overlap_20k <- integer()
#Fishers_exact_test <- integer()

for (row in 1:nrow(cutoff_combinations)){
  
  
  # Give number of remaining A1 DEGs
  A1_DEGs_curr <- GSE63127_CS_NS[GSE63127_CS_NS$adj.P.Val<cutoff_combinations$FDR_A1[row]
                                    & abs(GSE63127_CS_NS$logFC)>cutoff_combinations$log2FC_A1[row],]
  A1_DEGs <- c(A1_DEGs, nrow(A1_DEGs_curr))
  
  # Give number of remaining TE DEGs
  TE_DEGs_curr <- TCGA_LUSC_DEG[TCGA_LUSC_DEG$FDR<cutoff_combinations$FDR_TE[row]
                              & abs(TCGA_LUSC_DEG$log2FC)>cutoff_combinations$log2FC_TE[row],]
  TE_DEGs <- c(TE_DEGs, nrow(TE_DEGs_curr))
  
  # Give number of remaining TM DMGs
  TM_DMGs_curr <- TCGA_LUSC_DMG[TCGA_LUSC_DMG$FDR<cutoff_combinations$FDR_TM[row]
                               & abs(TCGA_LUSC_DMG$log2FC)>cutoff_combinations$log2FC_TM[row],]
  TM_DMGs <- c(TM_DMGs,nrow(TM_DMGs_curr))
  
  # Give number of "linked genes" resulting using my function
  linked_genes <- c(linked_genes, 
                    linked_gene_result
                    (A1_DEGs_curr, TE_DEGs_curr, TM_DMGs_curr,
                    cutoff_combinations$log2FC_A1[row], cutoff_combinations$FDR_A1[row],
                    cutoff_combinations$log2FC_TE[row], cutoff_combinations$FDR_TE[row],
                    cutoff_combinations$log2FC_TM[row], cutoff_combinations$FDR_TM[row]))
  
  # Give expected number of overlapping genes given random draws from 20k genes
  expected_overlap_20k_curr <- nrow(A1_DEGs_curr)/20000*nrow(TE_DEGs_curr)/20000*nrow(TM_DMGs_curr)
  expected_overlap_20k <- c(expected_overlap_20k, expected_overlap_20k_curr)
  
}

cutoff_combinations$A1_DEGs <- A1_DEGs
cutoff_combinations$TE_DEGs <- TE_DEGs
cutoff_combinations$TM_DMGs <- TM_DMGs
cutoff_combinations$linked_genes <- linked_genes
cutoff_combinations$expected_overlap_20k <- expected_overlap_20k

# Filter to acceptable ranges for the 3 datasets, and then pick a combination that optimizes the Fisher's exact test result
# cutoff_combinations_filtered <- cutoff_combinations %>%
#   filter(A1_DEGs %in% 1000:7500) %>%
#   filter(TE_DEGs %in% 4000:7500) %>%
#   filter(TM_DMGs %in% 4000:7500)

```


## 2025/02/10

Checking the TCGA-LUSC methylation data ONLY tumor samples comparing former and never smokers using the list of "linked genes" and "persistent linked genes". Will try to do this using first PCA and then a heatmap (hierarchical clustering).

## Read in TCGA-LUSC clinical data, filter out current smokers, order by smoking status
```{r}

library(TCGAbiolinks)
library(SummarizedExperiment)
library(dplyr)

## Read in TCGA-LUSC clinical data
query_clinical <- GDCquery(
  project = "TCGA-LUSC",
  data.category = "Clinical",
  data.type = "Clinical Supplement"
)

# Check the available files
query_clinical$results[[1]] %>% head()

## I see that some are not BCR XML files, so I will try to remove these
query_clinical$results[[1]] <- query_clinical$results[[1]] %>%
  filter(data_format == "BCR XML")

## Download data
GDCdownload(query_clinical)
clinical_data <- GDCprepare_clinic(query_clinical, clinical.info = "patient")

# Filter out current smokers (tobacco smoking history = 2) and former smokers with no specified duration (tsh = 5) and sort by tobacco smoking history
clinical_data_no_CS <- clinical_data %>%
  filter(tobacco_smoking_history != 2) %>% # remove current smokers
  filter(tobacco_smoking_history != 5) %>% # remove former smokers of unspecified duration
  arrange(., tobacco_smoking_history) %>%
  select(bcr_patient_barcode, tobacco_smoking_history)

# Get the patient barcodes in a form that can be used to filter the tumor samples
clinical_data_no_CS_sample_barcodes <- clinical_data_no_CS %>%
  mutate(bcr_patient_barcode = bcr_patient_barcode %>%
    gsub("$", ".01", .) %>%
    gsub("-", "\\.", .)
  )

## Read in the data for TCGA-LUSC methylation
LUSC_methylation_hm450_tumor <- read.table("../4_TCGA_data/TCGA_LUSC/data_methylation_hm450_LUSC.txt", header=TRUE, fill=TRUE)

# Get the barcodes existing in rownames of the sample info
clinical_data_no_CS_sample_barcodes_in_methyl <- clinical_data_no_CS_sample_barcodes[clinical_data_no_CS_sample_barcodes$bcr_patient_barcode %in% colnames(LUSC_methylation_hm450_tumor),] 

```


## Deal with duplicate genes and make them the rownames of the methylation table
```{r}
### 3.2.2 Handling duplicate genes
library(dplyr)

# Step 1: Identify rows with duplicate Gene values
duplicates <- LUSC_methylation_hm450_tumor %>%
  group_by(Hugo_Symbol) %>%
  filter(n() > 1)

# Step 2: Remove the row with the smallest row sum for each duplicate gene
rows_to_remove <- duplicates %>%
  rowwise() %>%
  mutate(row_sum = sum(c_across(where(is.numeric)))) %>%  # Compute row sum for numeric columns
  group_by(Hugo_Symbol) %>%
  slice_min(row_sum, with_ties = FALSE) %>%  # Select the row with the smallest sum
  ungroup()

# Step 3: Remove these rows from the original dataframe
LUSC_methylation_hm450_tumor <- anti_join(LUSC_methylation_hm450_tumor, rows_to_remove, by = colnames(LUSC_methylation_hm450_tumor))

# Now make the gene names into row names
rownames(LUSC_methylation_hm450_tumor) <- LUSC_methylation_hm450_tumor$Hugo_Symbol

```



## PCAs of the TCGA-LUSC methylation NS FS using genes of interest
```{r}

# Make gene names as row names
rownames(LUSC_methylation_hm450_tumor) <- LUSC_methylation_hm450_tumor$Hugo_Symbol

## Filter to former and never smokers
LUSC_methylation_hm450_tumor_FS_NS <- LUSC_methylation_hm450_tumor %>%
  select(clinical_data_no_CS_sample_barcodes_in_methyl$bcr_patient_barcode)

## Filter to just "Linked genes"
LUSC_methylation_hm450_tumor_FS_NS_linked_genes <- LUSC_methylation_hm450_tumor_FS_NS %>%
  filter(rownames(.) %in% A1_TE_TM_linked_genes$Gene)

## Plot a PCA based on these genes, using color to represent smoking status

## PCA to check for tumor-normal separation with outlier removed
colz <- clinical_data_no_CS_sample_barcodes_in_methyl$tobacco_smoking_history # Get color values from smoking history
plotMDS(LUSC_methylation_hm450_tumor_FS_NS_linked_genes,
        gene.selection = "common",
        main = "PCA for TCGA-LUSC former and never smokers using linked genes",
        col = colz,
        pch = 1
)

legend("topright", legend = c("1 = Lifelong Non-smokers", "3 = Current reformed smokers for >15 years", "4 = Current reformed smokers for ≤15 years"), 
       fill = unique(colz), 
       title = "Smoking status")


## Plot a heatmap with hierarchical clustering to achieve the same thing
```

### Checking the same thing using "persistent linked genes"

```{r}
## Filter to just "Linked genes"
LUSC_methylation_hm450_tumor_FS_NS_persistent_linked_genes <- LUSC_methylation_hm450_tumor_FS_NS %>%
  filter(rownames(.) %in% A1_TE_TM_A2_persistent_linked_genes$Gene)

## Plot a PCA based on these genes, using color to represent smoking status

## PCA to check for tumor-normal separation with outlier removed
colz <- clinical_data_no_CS_sample_barcodes_in_methyl$tobacco_smoking_history # Get color values from smoking history
plotMDS(LUSC_methylation_hm450_tumor_FS_NS_persistent_linked_genes,
        gene.selection = "common",
        main = "PCA for TCGA-LUSC former and never smokers using persistent linked genes",
        col = colz,
        pch = 1
)

legend("topright", legend = c("1 = Lifelong Non-smokers", "3 = Current reformed smokers for >15 years", "4 = Current reformed smokers for ≤15 years"), 
       fill = unique(colz), 
       title = "Smoking status")


## Plot a heatmap with hierarchical clustering to achieve the same thing
```





### 2025/02/10
I want to check the results I get from the TCGA T vs NM differential analyses against anything I can find in literature

```{r}
library(ggvenn)
library(ggplot2)

## Read in the table of differentially expressed genes from the publication
LUAD_LUSC_FDRs_26292924 <- read.table('../4_TCGA_data/LUAD_LUSC_FDRs_26292924.txt', header = T)

# Extract just the LUSC info
LUSC_FDRs_26292924 <- LUAD_LUSC_FDRs_26292924[,c('LUSC', 'FDR_LUSC')]
colnames(LUSC_FDRs_26292924) <- c('Gene', 'FDR_LUSC')

# First just check the genes in common
LUSC_genes_26292924 <- LUSC_FDRs_26292924$Gene
LUSC_genes_limma <- TCGA_LUSC_DEG$Gene

gene_lists <- list(
  LUSC_genes_26292924 = LUSC_genes_26292924,
  LUSC_genes_limma = LUSC_genes_limma
)

ggvenn(gene_lists, 
       fill_color = c("skyblue", "lightgreen"), 
       stroke_size = 0.5, 
       set_name_size = 5)

## That's odd that only 50% of gene names are matching here...

## Anyway, I want to try comparing the FDRs here
LUSC_genes_comparing_26292924_limma <- inner_join(LUSC_FDRs_26292924, TCGA_LUSC_DEG, by = 'Gene')
colnames(LUSC_genes_comparing_26292924_limma) <- c("Gene", "FDR_LUSC_26292924","log2FC_LUSC_limma", "FDR_LUSC_limma")

ggplot(LUSC_genes_comparing_26292924_limma, aes(x=FDR_LUSC_26292924, y=FDR_LUSC_limma)) +
    geom_point(shape=1) +      # Use hollow circles
    coord_trans(y ='log10', x='log10')

## NO correlation AT ALL!!!


## Now I want to see how many are "significant" for each
nrow(LUSC_genes_comparing_26292924_limma[LUSC_genes_comparing_26292924_limma$FDR_LUSC_26292924<0.05,])
nrow(LUSC_genes_comparing_26292924_limma[LUSC_genes_comparing_26292924_limma$FDR_LUSC_limma<0.05,])

# Seems maybe mine is not giving a stringent enough result?



```


